instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/EmbeddedConfigFile.java

```diff
@@ -0,0 +1,143 @@
+
+    Map<String, String> props = new HashMap<>();
+    for (PropertySource<?> propertySource : propertySources) {
+      if (propertySource instanceof EnumerablePropertySource<?>) {
+        for (String propertyName :
+            ((EnumerablePropertySource<?>) propertySource).getPropertyNames()) {
+          if (propertyName.startsWith("otel.")) {
+            String property = environment.getProperty(propertyName);
+            if (Objects.equals(property, "")) {
+              property = null; // spring returns empty string for yaml null
+            }
+            props.put(propertyName.substring("otel.".length()), property);
+          }
+        }
+      }
+    }
+
+    if (props.isEmpty()) {
+      throw new IllegalStateException("No application.yaml file found.");
+    }
+    return props;
+  }
+
+  static OpenTelemetryConfigurationModel convertToOpenTelemetryConfigurationModel(
+      Map<String, String> flatProps) {
+    Map<String, Object> nested = convertFlatPropsToNested(flatProps);
+
+    return getObjectMapper().convertValue(nested, OpenTelemetryConfigurationModel.class);
+  }
+
+  static ObjectMapper getObjectMapper() {
+    try {
+      Field field = DeclarativeConfiguration.class.getDeclaredField("MAPPER");
+      field.setAccessible(true);
+      return (ObjectMapper) field.get(null);
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      throw new DeclarativeConfigException("Failed to access ObjectMapper", e);
+    }
+  }
+
+  /**
+   * Convert flat property map to nested structure. e.g. "otel.instrumentation.java.list[0]" = "one"
+   * and "otel.instrumentation.java.list[1]" = "two" becomes: {otel: {instrumentation: {java: {list:
+   * ["one", "two"]}}}}
+   */
+  @SuppressWarnings("unchecked")
+  static Map<String, Object> convertFlatPropsToNested(Map<String, String> flatProps) {
+    Map<String, Object> result = new HashMap<>();
+
+    for (Map.Entry<String, String> entry : flatProps.entrySet()) {
+      String key = entry.getKey();
+      String value = entry.getValue();
+
+      // Split the key by dots
+      String[] parts = key.split("\\.");
+      Map<String, Object> current = result;
+
+      for (int i = 0; i < parts.length; i++) {
+        String part = parts[i];
+        boolean isLast = (i == parts.length - 1);
+
+        // Check if this part contains an array index like "list[0]"
+        Matcher matcher = ARRAY_PATTERN.matcher(part);
+        if (matcher.matches()) {
+          String arrayName = matcher.group(1);
+          int index = Integer.parseInt(matcher.group(2));
+
+          // Get or create the list
+          if (!current.containsKey(arrayName)) {
+            current.put(arrayName, new ArrayList<>());
+          }
+          List<Object> list = (List<Object>) current.get(arrayName);
+
+          // Ensure the list is large enough
+          while (list.size() <= index) {
+            list.add(null);
+          }
```

Comment:
ArrayList has `ensureCapacity` method that might simplify this

Reply:
no, this doesn't work, as `ensureCapacity` doesn't set the size